### PR TITLE
feature: debug_traceCall RPC support

### PR DESCRIPTION
**Pull request by XinyuCRO**
## Description

This PR introduces the [debug_traceCall](https://www.quicknode.com/docs/ethereum/debug_traceCall) RPC method

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: tests/integration_tests/test_tracers.py [hunk 1]
```diff
@@ -1,28 +1,31 @@
+import pytest
```
Comment (yihuang) [inline tests/integration_tests/test_tracers.py]:
```suggestion
```

File: rpc/backend/tracing.go [hunk 2]
```diff
@@ -211,3 +211,58 @@ func (b *Backend) TraceBlock(height rpctypes.BlockNumber,
...
+		GasCap:          b.RPCGasCap(),
+		ProposerAddress: sdk.ConsAddress(blk.Block.ProposerAddress),
+		BlockNumber:     blk.Block.Height,
+		BlockHash:       common.Bytes2Hex(blk.BlockID.Hash),
+		BlockTime:       blk.Block.Time,
+		ChainId:         b.chainID.Int64(),
+	}
+
+	if config != nil {
+		traceCallRequest.TraceConfig = config
```
Comment (mmsqe) [inline rpc/backend/tracing.go]: we might add `b.convertConfig(config)` here like https://github.com/crypto-org-chain/ethermint/commit/ccb83e2e068878551bc7bd704617bdaba72f1358

**Review by yihuang — APPROVED** [review]

**Comment by yihuang** [comment]
@XinyuCRO can you check the "allow maintainers to edit" so we can directly fix some small issues.

File: rpc/backend/backend.go [hunk 3]
```diff
@@ -139,6 +139,7 @@ type EVMBackend interface {
 	// Tracing
 	TraceTransaction(hash common.Hash, config *rpctypes.TraceConfig) (interface{}, error)
 	TraceBlock(height rpctypes.BlockNumber, config *rpctypes.TraceConfig, block *tmrpctypes.ResultBlock) ([]*evmtypes.TxTraceResult, error)
+	TraceCall(args evmtypes.TransactionArgs, blockNr rpctypes.BlockNumberOrHash, config *evmtypes.TraceConfig) (interface{}, error)
```
Comment (mmsqe) [inline rpc/backend/backend.go]:
```suggestion
	TraceCall(args evmtypes.TransactionArgs, blockNr rpctypes.BlockNumberOrHash, config *rpctypes.TraceConfig) (interface{}, error)
```

File: rpc/backend/tracing.go [hunk 4]
```diff
@@ -211,3 +211,58 @@ func (b *Backend) TraceBlock(height rpctypes.BlockNumber,
 
 	return decodedResults, nil
 }
+
+// TraceCall returns the structured logs created during the execution of EVM call
+// and returns them as a JSON object.
+func (b *Backend) TraceCall(
+	args evmtypes.TransactionArgs, blockNrOrHash rpctypes.BlockNumberOrHash, config *evmtypes.TraceConfig,
```
Comment (mmsqe) [inline rpc/backend/tracing.go]:
```suggestion
	args evmtypes.TransactionArgs, blockNrOrHash rpctypes.BlockNumberOrHash, config *rpctypes.TraceConfig,
```

File: rpc/backend/tracing.go (hunk 2 above)
Comment (mmsqe) [inline rpc/backend/tracing.go]:
```suggestion
		traceCallRequest.TraceConfig = b.convertConfig(config)
```

File: rpc/namespaces/ethereum/debug/api.go [hunk 5]
```diff
@@ -117,6 +117,17 @@ func (a *API) TraceBlockByHash(hash common.Hash, config *rpctypes.TraceConfig) (
 	return a.backend.TraceBlock(rpctypes.BlockNumber(resBlock.Block.Height), config, resBlock)
 }
 
+// TraceCall returns the structured logs created during the execution of EVM call
+// and returns them as a JSON object.
+func (a *API) TraceCall(
+	args evmtypes.TransactionArgs,
+	blockNrOrHash rpctypes.BlockNumberOrHash,
+	config *evmtypes.TraceConfig,
```
Comment (mmsqe) [inline rpc/namespaces/ethereum/debug/api.go]:
```suggestion
	config *rpctypes.TraceConfig,
```

File: tests/integration_tests/test_tracers.py [hunk 6]
```diff
@@ -83,3 +87,336 @@ def process(w3):
...
+    assert tx_res["result"] == EXPECTED_STRUCT_TRACER, ""
+
+    tx_res = eth_rpc.make_request(
+        "debug_traceCall", [tx, "latest", {"tracer": "callTracer"}]
+    )
+    assert tx_res["result"] == EXPECTED_CALLTRACERS, ""
+
+    tx_res = eth_rpc.make_request(
+        "debug_traceCall",
+        [tx, "latest", {"tracer": "callTracer", "tracerConfig": "{'onlyTopCall':True}"}],
```
Comment (mmsqe) [inline tests/integration_tests/test_tracers.py]:
```suggestion
        [tx, "latest", {"tracer": "callTracer", "tracerConfig": {"onlyTopCall": True}}],
```

**Review by mmsqe — COMMENTED** [review]
just config nitpick, others LGTM

File: x/evm/keeper/state_transition.go [hunk 7]
```diff
@@ -358,18 +364,21 @@ func (k *Keeper) ApplyMessageWithConfig(ctx sdk.Context,
 
 	evm := k.NewEVM(ctx, msg, cfg, tracer, stateDB)
 	leftoverGas := msg.GasLimit
+	sender := vm.AccountRef(msg.From)
 	// Allow the tracer captures the tx level events, mainly the gas consumption.
 	vmCfg := evm.Config
 	if vmCfg.Tracer != nil {
+		stateDB.SubBalance(sender.Address(), new(big.Int).Mul(msg.GasPrice, new(big.Int).SetUint64(msg.GasLimit)))
```
Comment (yihuang) [inline x/evm/keeper/state_transition.go]: if the node set config `evm.tracer`, the consensus will also have non nil `Tracer`.

Comment (XinyuCRO) [reply x/evm/keeper/state_transition.go]: ok, I thought only tracing RPC calls would have the Tracer parameter, I'll add a separate parameter to indicate whether it's a debug tracer call

**Review by yihuang — APPROVED** [review]

**Comment by XinyuCRO** [comment]
I create another PR https://github.com/crypto-org-chain/ethermint/pull/390 to resolve this commit sign issue, this one can be closed

**Pull request closed without merging**